### PR TITLE
Pull missing transactions and sync unknown transactions as separate processes

### PR DESCRIPTION
**Pull request by a-urth**
- pull missing transactions directly by id within one request (assumption that it always will be only a few of those);
 - sync mempool (index) transactions through grpc stream using bloom filter as source of known transactions;
 - remove error as return value from ledger's `AddTransaction` since it never returns an actual error

TODO:
 - ~~move transaction syncing parameters to config~~

### CODE REVIEW

**Comment by iwasaki-kenta** [comment]
Do logs come up indicating that transactions pulled which were missing are not being added because of height difference, etc?

**Comment by a-urth** [comment]
There are no logs for that, since this validation results in just `return` in `transactions.add`. I added prints, but didn't reproduce it yet, that why it was and assumption.

**Comment by rkeene** [comment]
I'm reviewing this now

File: ledger.go [hunk 1]
```diff
@@ -238,12 +243,14 @@ func (l *Ledger) Close() {
 // invalid or fails any validation checks, an error is returned. No error
```
Comment (rkeene) [inline ledger.go]: The comment for this function needs to be updated to indicate that it accepts many transactions and remove references to failing

File: ledger.go [hunk 2]
```diff
@@ -380,56 +383,186 @@ func (l *Ledger) PullTransactions() {
...
-						responseChan <- nil
+				res, err := stream.Recv()
+				if err != nil {
+					logger.Error().Err(err).Msg("failed to receive sync transactions header")
+					return
+				}
+
+				count := res.GetTransactionsNum()
+				if count == 0 || count > maxSize {
+					return
```
Comment (rkeene) [inline ledger.go]: What do you think about logging the case where too many transactions are received (and thus we don't accept any?)

```suggestion
					if count != 0 {
						logger.Debug().Msg("Bad number of transactions would be received").Uint64("count", count)
					}
					return
```

File: ledger.go [hunk 3]
```diff
@@ -380,56 +383,186 @@ func (l *Ledger) PullTransactions() {
...
+				}
 
+				if count > 0 {
```
Comment (rkeene) [inline ledger.go]: This should always be true at this point because of the check right above it

File: ledger.go [hunk 4]
```diff
@@ -380,56 +383,186 @@ func (l *Ledger) PullTransactions() {
...
-					cancel()
 
-					responseChan <- batch
+					res, err := stream.Recv()
+					if err != nil {
+						logger.Error().Err(err).Msg("failed to receive sync transactions header")
+						return
+					}
+
+					transactions := res.GetTransactions()
```
Comment (rkeene) [inline ledger.go]: What do you think about calling this "`rawTransactions`" or "`transactionsRaw`" or "`txsRaw`" and the "`txs`" below as "`transactions`" to make it a bit more unambiguous ?

File: ledger.go [hunk 5]
```diff
@@ -380,56 +383,186 @@ func (l *Ledger) PullTransactions() {
 			select {
```
Comment (rkeene) [inline ledger.go]: This whole `select` section can be removed since the very first thing the top of the loop does is this

File: ledger.go [hunk 6]
```diff
@@ -380,56 +383,186 @@ func (l *Ledger) PullTransactions() {
...
+		case <-l.sync:
+			return
+		case <-time.After(100 * time.Millisecond):
+		}
+
+		snowballK := conf.GetSnowballK()
+
+		peers, err := SelectPeers(l.client.ClosestPeers(), snowballK)
+		if err != nil {
+			select {
```
Comment (rkeene) [inline ledger.go]: This whole `select` section can be removed (same as `SyncTransactions()`)

File: transactions.go [hunk 7]
```diff
@@ -59,10 +59,14 @@ func (t *Transactions) BatchAdd(block BlockID, transactions ...Transaction) {
 
 func (t *Transactions) add(block BlockID, tx Transaction) {
 	if t.height >= tx.Block+uint64(conf.GetPruningLimit()) {
+		fmt.Println(">>>>>>>>>>>> why??????", t.height, tx.Block)
```
Comment (rkeene) [inline transactions.go]: We'll leave this debugging in for now as we track it down

File: transactions.go [hunk 8]
```diff
@@ -59,10 +59,14 @@ func (t *Transactions) BatchAdd(block BlockID, transactions ...Transaction) {
...
 		return
 	}
 
 	if _, exists := t.buffer[tx.ID]; exists {
+		if _, ok := t.missing[tx.ID]; ok {
+			fmt.Println(">>>>>>>>>>>> WHOAAAAAAAAAAAAAA")
```
Comment (rkeene) [inline transactions.go]: We'll leave this debugging in for now as we ensure it's fixed, but maybe this should be a logged debug message in the future ?

**Comment by a-urth** [comment]
@rkeene did changes according to all of your comments

**Pull request merged**